`ot_usbdev`: add initial host protocol and bus handling (#238)
lowRISC/qemu: docs/opentitan/usbdev.md

  will be completely independent of the presence of a USB host.
- `vbus-override=off`: in this mode, the `vbus_on` and `vbus_off` commands control a virtual "VBUS
  enable" gate. The VBUS sense pin will be reported as level high if and only if the VBUS enable
  gate is enabled **and** a USB host has enabled VBUS (see [host simulation](#host-simulation))
  By default, the VBUS gate is enabled.

## Host simulation

The QEMU USBDEV driver only simulates a USB device controller (UDC) which requires the presence
of a USB host. A (virtual) host exchanges messages with the UDC over the [`udev-host` chardev](#usbdev-host-chardev).

> This binary protocol is somewhat similar to the [USB/IP](https://docs.kernel.org/usb/usbip_protocol.html)

Reviewer: nit pick: might be more readable to use a "footnote" link

This binary protocol is somewhat similar to the [USB/IP][usbip]
protocol but lower level. This protocol is also different from the [USB network redirection (a.k.a usbredir)][usbredir]
protocol supported by QEMU.

[usbip]: (https://docs.kernel.org/usb/usbip_protocol.html)
[usbredir]: (https://gitlab.freedesktop.org/spice/usbredir/-/blob/main/docs/usb-redirection-protocol.md)

### Rationale for a different protocol

Both the USB/IP and usbredir protocols are too high-level for the purpose of emulating and testing a low-level
UDC driver. For example, both protocols require that the device be fully enumerated and configured before the
host is even made aware of its presence. In contrast, we want to be able to fully emulate the enumeration sequence.
Another big difference is bus management: USB/IP does not support bus resets and usbredir does not support resume/suspend,
neither model VBUS. This is a side-effect of the intended usecase of these protocols: connect a real USB device
to a virtual USB host. However this new protocol should be low-level enough that it is possible to implement a

> bridge from either usbredir or USB/IP to this protocol (while loosing some features).

Reviewer 2: Grammar nit:
```suggestion
and neither models VBUS. This is a side-effect of the intended use case of these protocols: to connect a real USB device
to a virtual USB host. However, this new protocol should be low-level enough that it is possible to implement a
bridge from either usbredir or USB/IP to this protocol (while losing some features).
```

### High-level overview

This protocol does not specify which of the device or the host should the client/guest.
On connection, the client must send a [`Hello` command](#hello-command), to which the server

> must respond by another `Hello` command. After that, messages are exchanged asychronously.

Reviewer 2: ```suggestion
This protocol does not specify which of the device or the host should be the client/guest.
On connection, the client must send a [`Hello` command](#hello-command), to which the server
must respond with another `Hello` command. After that, messages are exchanged asynchronously.
```

#### Bus states

The protocol reflects the low-level details of the USB bus and provides independent
handling of VBUS (controlled by the host) and connection (controlled by the device).
The host can turn VBUS on and off at any point. When VBUS is on, the device can freely

> connect or disconnect by asychronously sending a message to the host. As on a real bus,

Reviewer 2: (occurs a couple more times)
```suggestion
connect or disconnect by asynchronously sending a message to the host. As on a real bus,
```

turning off VBUS disconnects the device. VBUS is assumed to be initially off.

It is an error for the device to send a message when VBUS is not on and the host must ignore such

> messages. However due to the asychronous nature of the protocol, it is possible for the host to

Reviewer 2: ```suggestion
messages. However due to the asynchronous nature of the protocol, it is possible for the host to
```

| VBUS On     |     2     | Host to device | Turn VBUS on | [Bus commands](#bus-commands) |
| VBUS Off    |     3     | Host to device | Turn VBUS off | [Bus commands](#bus-commands) |
| Connect     |     4     | Device to host | Report device connection | [Bus commands](#bus-commands) |
| Disconnect  |     5     | Device to host | Report device disconnection | [Bus commands](#bus-commands) |
| Reset       |     6     | Host to device | Reset the device | [Bus commands](#bus-commands) |
| Resume      |     7     | Host to device | Resume the device | [Bus commands](#bus-commands) |
| Suspend     |     8     | Host to device | Suspend the device | [Bus commands](#bus-commands) |

**TODO** Add transfer commands

#### Hello command

The ID of this command must be 0.
The payload of this command is defined as follows.

| **Field** | **Offset** | **Size** | **Value** |
| --------- | ---------- | -------- | --------- |
| Magic value |      0   |     4    | String `UDCX` |
| Major version |   4    |     2    |      1    |
| Minor version |   6    |     2    |     0     |

#### Bus commands

> These commands (VBUS On/Off, (Dis)connection, Reset, Suspemd/Resume) do not have any payload.

Reviewer 2: ```suggestion
These commands (VBUS On/Off, (Dis)connection, Reset, Suspend/Resume) do not have any payload.
```
